Add custom effect manager

cirrahn opened the pull request:
This allows a GM user to customize their FX.

Features:

- May define macros for items by "name" or "Lancer ID"
- Macros overwrite module defaults
- Extensive drag-drop support
- Folder system
- "Execute Macro" preview button
- "Export/Import" buttons to transfer state between worlds
- Effects Manger respects system theme choice

---

Adds a new button in the settings:

![00-settings](https://github.com/user-attachments/assets/446cf387-2e7d-463c-9cbe-9bc393becae4)

Effects can be created and existing macros selected:

[create_effect.webm](https://github.com/user-attachments/assets/67f252a4-f3d1-4d4c-b8c3-fc9298de896d)

Effects can be organized into folders:

[folders.webm](https://github.com/user-attachments/assets/0ead8d55-371f-4179-a0ca-4fffa90b4db9)

Effects and folders can be deleted, with additional "also delete contents" option for folders:

[delete.webm](https://github.com/user-attachments/assets/af0a4c82-661b-4cc2-94d1-a9f8e07bd3fc)

Everything can be exported/imported to backup or move between worlds:

[export-import.webm](https://github.com/user-attachments/assets/f0c729ef-e1d4-4d24-ba18-6d79eaba4188)

Items can be dropped to:

- A folder, to create a new effect in that folder
- The app, to create a new effect
- An existing effect, to set the LID (if available) or name (if LID unavailable) for that effect

[drag-drop-items.webm](https://github.com/user-attachments/assets/f3d25394-e290-4779-b64f-aa334674a46f)

Sheet items may also be dropped:

[drag-drop-sheet-items.webm](https://github.com/user-attachments/assets/8a6ee36e-1210-4fd0-9965-814d078c5373)

Macros can be dropped to:

- A folder, to create a new effect in that folder
- The app, to create a new effect
- An existing effect, to set the macro for that effect

[drag-drop-custom-macro.webm](https://github.com/user-attachments/assets/249627c4-1818-44b6-9e44-65af097f6eeb)

Macros can be previewed (executed) from the UI. 

Not pictured: if no target selected, or the player does not have a token, a warning is shown and the macro is not executed. This re-uses the existing token resolution logic.

[execute-macro.webm](https://github.com/user-attachments/assets/7798f6d1-fc5b-4969-b5ba-7d3081ead3ef)

Custom effects playing when items are used:

[custom-effects.webm](https://github.com/user-attachments/assets/28be91ae-ec13-4cc4-bcac-4b91c519f2c4)

Custom effects are stored as a GM world setting. This is synced if there are multiple GMs:

[multi-gm-sync.webm](https://github.com/user-attachments/assets/dce8948b-60a7-41e8-b398-cb503bec182d)

The UI is "responsive":

[responsive.webm](https://github.com/user-attachments/assets/0a3693fc-fa28-4096-aae7-bed7f01b91ac)

The UI respects the system theme choice of the user:

[themes.webm](https://github.com/user-attachments/assets/c0327ba7-9f19-4c1a-bfd0-8ec30e789d37)

---

Testing is as simple as checking out the branch and using the module. No additional build steps required.

---

Discussion points:

- [x] Should there be a confirmation dialogue when deleting an effect and/or folder?
- [x] "How to" guide - Foundry Tour?
- [ ] Others?

cirrahn commented on css/effects-manager.css:
```diff
@@ -0,0 +1,68 @@
+.lwfx-effects-manager {
```
most variables (`--var(...)`) are from the Lancer system. This provides the theme support

cirrahn commented on css/helpers.css:
```diff
@@ -0,0 +1,112 @@
+/* TODO(v12) remove some of these in favor of v2 styling */
```
Foundry v12 has more and better utility classes. Sadly v11 does not so I made the ones I needed instead.

cirrahn commented on scripts/api.js:
```diff
@@ -11,29 +11,6 @@ class ModuleApi {
         return volume * game.settings.get(MODULE_ID, SETTING_VOLUME);
     }
 
-    static getMacroVariables(macro) {
```
Moved to shared `utils.js` so it can be used by the "Execute Macro" button in the UI

cirrahn commented on scripts/effectManager/app.js:
```diff
@@ -0,0 +1,509 @@
+            id: `${MODULE_ID}-effects-manager`,
+            template: `modules/${MODULE_ID}/templates/effectManager/effect-manager.hbs`,
+            title: game.i18n.localize(`${MODULE_ID}.effectManager.app.Effects Manager`),
+            width: 800,
+            height: 800,
```
800x800 is the dimensions of the Lancer mech sheet. It felt comfortable!

cirrahn commented on scripts/effectManager/models.js:
```diff
@@ -0,0 +1,93 @@
+/* -------------------------------------------- */
+
+/**
+ * Notes:
```
This is not ideal, but it works. In v12 the implementation does not change, but they are committed to expanding datamodel/field support so maybe in v13+ this will need a revisit!

cirrahn commented on scripts/utils.js:
```diff
@@ -27,3 +30,26 @@ export function getUniquePoints(points) {
     });
 }
+
+export function getMacroVariables(macro = null) {
```
Moved from `api.js`

Z3nner commented:
I can absolutely see myself deleting fx by accident when I attempt to test it, but I can also see occasion where someone wants to delete a lot of fx, and repetitive confirmation dialogs would be annoying.
Here is what I would propose.
For folders
- move the new effect button to the left side, right of the collapse folder button. 
- Keep the delete functionality as is but add text to the button: 'Remove Folder' (the instant tooltip is clear and obvious and I like the functionality of removing entire sets at once.)

For individual effects

- replace the delete button with a checkbox, and add a Delete Selected button to the top menu, that pops up a confirmation dialog to confirm the user wants to delete all selected effects.


Can it handle non-weapon gear like mech systems, or core powers?
depending on the answer to that, when there are no effects loaded at all, as in, the first time someone opens the manager, would it be more clear to state something like, "Drag and drop a weapon or system, or the macro you wish to pair with that weapon or system to begin a new pairing."

cirrahn commented:
> For folders ...

Like this? (Ignoring bad tooltip casing)

![image](https://github.com/user-attachments/assets/81d36ef2-276b-4daf-b185-b5c96d932c9f)

> For individual effects ...

Sounds good. I will position the checkbox between the grab handle and the "MODE" box, and will make the grab handle invisible until the row is hovered.

---

> Can it handle non-weapon gear like mech systems, or core powers?

Any "Foundry Item" works, which includes systems. It does not include core powers -- good catch. Have fixed that.

---

> the first time someone opens the manager

Good point. It could be a ["Tour"](https://foundryvtt.com/article/tours/)!

cirrahn commented:
Another point which is buried in the code: duplicate handling.

Currently: when a custom FX plays, if there are multiple matches (2+ with the same LID, or 2+ with the same name) for the action, a warning notification is shown.

Should there be more proactive handling? A warning display on the row in the manager? Thinking a "⚠️" icon with a tooltip.

Z3nner commented:
I think that's a good solution. Could there be a tooltip that pops up with that icon hovered that says duplicate effect?

cirrahn commented:
PR updated.

Multi-select, select-all, and multi-delete with confirmation:

[multi-delete.webm](https://github.com/user-attachments/assets/87df9aea-89ef-41af-af7b-ebd3fe8bf187)

Hidden drag grabs:

[drag-grab-update.webm](https://github.com/user-attachments/assets/5286a1fb-876e-4e48-a07c-ebaf5ff438b5)

Duplicate warnings:

[duplicate-handling.webm](https://github.com/user-attachments/assets/3e70375b-dc0b-4cff-b5eb-40705867cf3c)

Tour:

[tour.webm](https://github.com/user-attachments/assets/646ee7cd-e6e0-45ad-9cbf-4ec32900ecba)

---

The tour also starts when the user opens the app for the first time. Unfortunately, there is a Foundry bug/race condition when hovering an element with a tooltip and starting a tour at the same time. This can prevent the tour tooltips from showing, and the user has to ESC to leave the tour without completing it. As it is possible to hit this when opening the manager from the settings menu, I have added a "Start Tour" button to the UI to allow the user to restart the tour at will.

cirrahn commented:
PR updated.

Added help text when no effects/folders are displayed. Also removed the "UNCATEGORIZED" section if no effects are displayed.

[hints.webm](https://github.com/user-attachments/assets/c3eb8702-511b-4a5b-9b57-25caad7bfc70)

The tour bug is now fixed too. I did some more investigation and it was:

- We click the settings button to open the app
- The app opens, triggering a "pointerleave" event on the button, as the app now occludes the button
- The tooltip manager queues up a "deactivate tooltip" task, on a timer
- Our app finishes rendering; our tour hook triggers, the tour start
- The tooltip manager's timer pops, and it deactivates our tour's tooltip.

"Restarting" the tour after a short delay prevents the issue and is invisible to the user.
